src/relay/ir/Relay.g4

```diff
@@ -0,0 +1,140 @@
+
+// non-negative ints
+// INT: '0' | [1-9] DIGIT* ; // no leading zeros
+INT: DIGIT+ ;
+fragment EXP: [eE] [+\-]? INT ; // \- since - means "range" inside [...]
+
+CNAME: ('_'|LETTER) ('_'|LETTER|DIGIT)* ;
+fragment LETTER: [a-zA-Z] ;
+fragment DIGIT: [0-9] ;
+
+// Parsing
+
+// a program is a list of options followed by either an expression or a list of definitions
+// TODO: is this really the desired outline? if there are defns should there be an expression at the end?
+prog: option* (expr | defn+) EOF ;
+
+option: 'set' ident BOOL_LIT ;
+
+expr
+  // operators
+  // todo: this precedence seems fishy
+  : '(' expr ')'                  # parens
+  | '-' expr                              # neg
+  | expr op=('*'|'/') expr                # binOp
+  | expr op=('+'|'-') expr                # binOp
+  | expr op=('<'|'>'|'<='|'>=') expr      # binOp
+  | expr op=('=='|'!=') expr              # binOp
+
+  // function definition and application
+  | expr '(' (expr (',' expr)*)? ')'      # call
+  | fun                                   # funExpr
+
+  // tuples and tensors
+  | '(' ')'                               # tuple
+  | '(' expr ',' ')'                      # tuple
+  | '(' expr (',' expr)+ ')'              # tuple
+  | '[' (expr (',' expr)*)? ']'           # tensor
+
+  // todo: precedence
+  | 'if' expr body 'else' body            # ifElse
+
+  // sequencingg
+  | 'let' MUT? ident (':' type_)? '=' expr ';' expr  # let
+  // sugar for let _ = expr; expr.
+  | expr ';' expr                         # seq
+
+  | expr '^'                              # readRef
+  // sugar for let _ = WriteRef(ident, expr); expr
+  | ident '=' expr ';' expr               # writeRef
+
+  | ident                         # identExpr
+  | scalar                        # scalarExpr
+  | expr '.' INT                  # project
+  | 'debug'                       # debug
+  ;
+
+fun:  'fn'        paramList '=>' type_? body ;
+defn: 'def' ident paramList '=>' type_? body ;
+
+paramList: '(' param (',' param)* ')' ;
+param: ident (':' type_)? ;
+
+type_
+  : '(' type_ ')'                           # parensType
+  | type_ op=('*'|'/') type_                # binOpType
+  | type_ op=('+'|'-') type_                # binOpType
+  | '(' ')'                                 # tupleType
+  | '(' type_ ',' ')'                       # tupleType
+  | '(' type_ (',' type_)+ ')'              # tupleType
+  | 'Tensor[' type_ ',' type_ ']'           # tensorType
+  | identType                               # identTypeType
+  | MUT '(' type_ ')'                       # refType
+  | baseType                                # baseTypeType
+  | type_ '.' INT                           # projectType
+  | INT                                     # dimLitType
+  | '_'                                     # placeholderType
+  ;
+
+identType: CNAME ;
+// int8, int16, int32, int64
+// uint8, uint16, uint32, uint64
+// float16, float32, float64
+// bool
+
+baseType
+  : 'int(' INT ',' INT ')' # intType
```

Comment:
Int can be any arbitrary number, is that expected?

Reply:
`INT` is really a natural number, which is what we want for projection and the inputs here, which represent bits and lanes. Maybe I should change the name to reflect that.

Reply:
I think the type checker should enforce constraints on this, it will yield a better error.

Reply:
What do you think about adding a type-level function syntax and moving this sort of parsing into the python parser?